modules/data-entry/src/main/java/io/uhndata/cards/PaginationServlet.java

```diff
@@ -145,18 +150,41 @@ private String createQuery(final SlingHttpServletRequest request)
         } catch (Exception e) {
             nodeType = request.getResource().getResourceType().replace('/', ':').replaceFirst("sHomepage$", "");
         }
-        final StringBuilder query =
-            // We select all child nodes of the homepage having the right type
-            new StringBuilder("select n.* from [").append(nodeType).append("] as n");
+        // We select all child nodes of the homepage having the right type
+        final StringBuilder query = new StringBuilder("select n.* from [").append(nodeType).append("] as n");
 
         // If child nodes are required for this query, also grab them
         final String[] filternames = request.getParameterValues("filternames");
-        query.append(createJoins(
-            request.getParameter("joinchildren"),
-            request.getParameterValues("filternames"),
-            request.getParameterValues("filterempty"),
-            request.getParameterValues("filternotempty")
-            ));
+        final String[] filterempty = request.getParameterValues("filterempty");
+        final String[] filternotempty = request.getParameterValues("filternotempty");
+        final String joinNodetype = request.getParameter("joinchildren");
+
+        // The map that stores questionnaires uuids with prefixes in array and maps it to the
+        // corresponding group of questions uuids with their prefixes (stored in a map)
+        // To be used later on the filter query construction
+        Map<String, String> questionnairesToPrefix = new HashMap<>();
+        Map<String, List<String>> questionnairesToFilters = new HashMap<>();
+        Map<String, String> filtersToPrefix = new HashMap<>();
+
+        // Add joints
+        if (StringUtils.isNotBlank(joinNodetype)) {
+            String sanitizednodetype = joinNodetype.replaceAll("[\\\\\\]]", "\\\\$0");
+            if (nodeType.equals(SUBJECT_IDENTIFIER)) {
+                // resolve all questions to questionnaires and group them by questionnaires
+                getQuestionnairesMaps(questionnairesToPrefix, questionnairesToFilters, filtersToPrefix, filternames,
+                    "child", session);
+                getQuestionnairesMaps(questionnairesToPrefix, questionnairesToFilters, null, filterempty,
+                    "empty", session);
+                getQuestionnairesMaps(questionnairesToPrefix, questionnairesToFilters, null, filternotempty,
```

Comment:
Why not this:

```suggestion
                getQuestionnairesMaps(questionnairesToPrefix, questionnairesToFilters, filtersToPrefix, filterempty,
                    "empty", session);
                getQuestionnairesMaps(questionnairesToPrefix, questionnairesToFilters, filtersToPrefix, filternotempty,
```

Without `filtersToPrefix`, there's a `NullPointerException` thrown at line 289.